### PR TITLE
Refactored setFunction and tests.

### PR DESCRIPTION
**Pull request by MrHIDEn**
I think I solved many things and I made setFunction universal one, also just using nerdamer can add a function too.

### CODE REVIEW

File: nerdamer.core.js [hunk 1]
```diff
@@ -165,14 +167,14 @@ var nerdamer = (function (imports) {
...
-            CB = Groups.CB = 7, // A symbol/expression composed of one or more variables through multiplication e.g. x*y
-            CP = Groups.CP = 8; // A symbol/expression composed of one variable and any other symbol or number x+1 or x+y
+    var N = Groups.N = 1; // A number
+    var P = Groups.P = 2; // A number with a rational power e.g. 2^(3/5).
+    var S = Groups.S = 3; // A single variable e.g. x.
+    var EX = Groups.EX = 4; // An exponential
+    var FN = Groups.FN = 5; // A function
+    var PL = Groups.PL = 6; // A symbol/expression having same name with different powers e.g. 1/x + x^2
+    var CB = Groups.CB = 7; // A symbol/expression composed of one or more variables through multiplication e.g. x*y
+    var CP = Groups.CP = 8; // A symbol/expression composed of one variable and any other symbol or number x+1 or x+y
```
Comment (MrHIDEn) [inline nerdamer.core.js:177]: Only tiny refactoring with by adding var for every variable.

File: nerdamer.core.js [hunk 2]
```diff
@@ -2715,6 +2725,7 @@ var nerdamer = (function (imports) {
 
             //Enable getting of expressions using the % so for example %1 should get the first expression
             if(typeof first_arg === 'string') {
+                //TODO Replace substr with slice, and test it
```
Comment (MrHIDEn) [inline nerdamer.core.js:2728]: should be slice. I couldnt test this place so I left it as it is.

File: nerdamer.core.js [hunk 3]
```diff
@@ -12465,6 +12473,33 @@ var nerdamer = (function (imports) {
         return result;
     };
 
+    /**
+     *
+     * @param {Boolean} asObject
+     * @param {String|String[]} option
+     * @returns {Array}
+     */
+    libExports.functions = function (asObject, option) {
```
Comment (MrHIDEn) [inline nerdamer.core.js:12482]: this returns list of user functions. I think it should be renamed to listFunctions or getFunctionList, maybe first one. and other things should be renamed too to keep the same pattern.

File: README.md [hunk 4]
```diff
@@ -293,48 +293,75 @@ console.log(e.text());
...
 console.log(answer);
 
 //result: 9
 ```
 
+Mode 2:
 Custom functions alternatively be set in following manner.
 
 ```javascript
-nerdamer('hyp(a, b) := sqrt(a^2 + b^2) ');
```
Comment (MrHIDEn) [inline README.md:320]: I fixed all places and now we can  declare functions with `=` nad `:=` as well (legacy)

**Comment by gunnarmein-ts** [comment]
Sorry I haven't processed this yet. I will look at it tomorrow.

**Pull request merged**